src/poetry/mixology/term.py

```diff
@@ -151,21 +151,45 @@ def difference(self, other: Term) -> Term | None:
         """
         return self.intersect(other.inverse)
 
+    @staticmethod
+    def _is_direct_reference(dependency: Dependency) -> bool:
+        return dependency.source_type in ["directory", "file", "url", "git"]
+
     def _compatible_dependency(self, other: Dependency) -> bool:
         compatible: bool = (
             self.dependency.is_root
             or other.is_root
             or other.is_same_package_as(self.dependency)
         )
+
+        if (
+            not compatible
+            and not self._is_direct_reference(self.dependency)
+            and self._is_direct_reference(other)
+        ):
+            # we do this here to indicate direct reference dependencies are compatible
+            # with NVR dependencies
+            return bool(
+                self.dependency.complete_name == other.complete_name
+                and self.dependency.constraint.allows_all(other.constraint)
```

Comment:
```suggestion
```
Checking the constraint is probably not necessary here because it is done later anyway. (`is_same_package_as()` does not consider the constraint, too.) Although it may not change anything, it seems more suitable not to do it here.

Reply:
I feel that that is incorrect. We do not want `demo (>=1.1.0)` to return compatible for `demo (1.0.0 git)`. I'd rather keep it in here and not rely on the later checks. Unless there is a pressing reason to remove it, I'd much rather keep it for correctness.

Reply:
But `demo (>=1.1.0)` is already compatible to `demo (<1.0)`. Removing the check allows the simplification in the comment below.

Reply:
At the moment, I think this method is not supposed to consider versions. That may seem weird at first glance, but at least it's consistent for now since versions are handled in the callers. It may even be wrong to handle it here that way because the handling depends on if the term is positive or negative.

Reply:
In that case we should just replace `is_same_package_as` with `self.dependency.complete_name == other.complete_name` right?

Reply:
This seems to not be the case with the current core. 

I have applied your suggested change for now. We can rework it if required after the next core release.

Reply:
`is_same_package_as()` (which does not consider versions) is required in some cases. `complete_name` is not enough:
If both dependencies are direct origin (or legacy?) dependencies, `is_same_package_as()` is required.
If both dependencies are PyPI dependencies, `is_same_package_as()` and `complete_name` are the same (does not matter which one to use).
If one dependency is a direct origin dependency and the other is a PyPI or legacy dependency, `complete_name` must be used instead of `is_same_package_as()`.

I'm only a bit unsure if two legacy dependencies from different sources or a PyPI and a legacy dependency are "compatible" or not. Perhaps, these are rare edge cases that can be revisited later.